Add implementation and contract change for serialization surrogate

khdang opened the pull request:
Fix #1931

khdang commented:
cc @SGuyGe @shmao @roncain

SGuyGe commented on src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs:
```diff
@@ -1846,6 +1846,66 @@ public static void DCS_DeserializeEmptyString()
     }
 
+    [Fact]
+    public static void DCS_DataContractSurrogate()
```
Rename it to DCS_MyPersonSurrogate() to be more specific?

khdang replied:
Done

shmao commented on src/System.Runtime.Serialization.Xml/ref/project.json:
```diff
@@ -3,7 +3,7 @@
     "System.Xml.ReaderWriter": "4.0.0",
     "System.IO": "4.0.0",
-    "System.Runtime.Serialization.Primitives": "4.0.10",
+    "System.Runtime.Serialization.Primitives": "4.1.0-beta-23423",
```
Do we want to reference this specific version of the package? Should this be "4.1.0-beta-*"?

khdang replied:
We shouldn't have reference to this specific version. I've updated as suggestion.

shmao commented on src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlObjectSerializerWriteContextComplex.cs:
```diff
@@ -21,13 +21,15 @@ public class XmlObjectSerializerWriteContextComplex : XmlObjectSerializerWriteCo
+        protected ISerializationSurrogateProvider SerializationSurrogateProvider;
         private SerializationMode _mode;
-
+        
```
nit: remove extra spaces.

shmao commented on src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlObjectSerializerReadContextComplex.cs:
```diff
@@ -26,12 +26,14 @@ internal class XmlObjectSerializerReadContextComplex : XmlObjectSerializerReadCo
         private bool _preserveObjectReferences;
         private SerializationMode _mode;
+        protected ISerializationSurrogateProvider SerializationSurrogateProvider;
 
```
Question: do we need this field to be "protected"?

khdang replied:
For now it doesn't need to be protected. I was thinking of adding the surrogate support for DCJS as well which would need this to be protected but we've agreed that can be done in a separated  change.

shmao commented on src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlObjectSerializerWriteContextComplex.cs:
```diff
@@ -21,13 +21,15 @@ public class XmlObjectSerializerWriteContextComplex : XmlObjectSerializerWriteCo
     internal class XmlObjectSerializerWriteContextComplex : XmlObjectSerializerWriteContext
 #endif
     {
+        protected ISerializationSurrogateProvider SerializationSurrogateProvider;
```
Question: do we need this field to be "protected"?

khdang replied:
Same comment as for XmlObjectSerializerReadContextComplex.cs

shmao commented:
:shipit: :+1:

SGuyGe commented:
LGTM

zhenlan commented:
LGTM

stephentoub commented:
@dotnet-bot test this please (https://github.com/dotnet/corefx/issues/4172)

stephentoub commented on src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSerializer.cs:
```diff
@@ -448,5 +458,23 @@ internal override Type GetDeserializeType()
+                if (obj == null)
+                    objType = Globals.TypeOfObject;
+                else
+                    objType = obj.GetType();
```
Nit: could be simplified to:
```C#
objType = obj != null ? obj.GetType() : Globals.TypeOfObject;
```

khdang replied:
Updated as suggestion

stephentoub commented on src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSurrogateCaller.cs:
```diff
@@ -0,0 +1,41 @@
+            Type dcType = surrogateProvider.GetSurrogateType(type);
+            if (dcType == null)
+                return type;
+            return dcType;
```
Nit: previous four lines could be simplified to:
```C#
return surrogateProvider.GetSurrogateType(type) ?? type;
```

khdang replied:
Updated as suggestion

stephentoub commented on src/System.Runtime.Serialization.Xml/ref/project.json:
```diff
@@ -3,7 +3,7 @@
     "System.Xml.ReaderWriter": "4.0.0",
     "System.IO": "4.0.0",
-    "System.Runtime.Serialization.Primitives": "4.0.10",
+    "System.Runtime.Serialization.Primitives": "4.1.0-beta-*",
```
It's ok to update this dependency?

khdang replied:
We've added a new interface to System.Runtime.Serialization.Primitives contract. From discussion we've agreed to take dependency of the new S.R.S.Primitives package.

stephentoub commented:
A few minor comments, but otherwise LGTM.

khdang commented:
@dotnet-bot test this please

khdang commented:
The CodeFormatter check failures are:
```
11:14:52 Formatting native file with command: clang-format -style=file d:\j\workspace\dotnet_corefx\native_code_format_check_prtest\src\Native\System.Native\pal_errno.cpp
11:14:52 File not style compliant, exiting: System.Native\pal_errno.cpp
11:14:52 
11:14:52 d:\j\workspace\dotnet_corefx\native_code_format_check_prtest>exit -1 
11:14:52 Build step 'Execute Windows batch command' marked build as failure
```
Hi @stephentoub My understanding is that we want to merge this soon so that it can be integrate to rc1 branch in time. Is there anything we can do about the failure? My first commit before addressing the comments and squash passed all the tests, should we go ahead and merge this change?

stephentoub commented:
You're not changing native code.  Don't worry about the Code Formatter Check.  If it passes the main leg with all the builds and tests, you're welcome to merge.  We're working on getting the code formatter leg fixed.
